cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs

```diff
@@ -198,6 +198,44 @@ runTransactionBuildCmd
         first TxCmdProposalError
           <$> readTxGovernanceActions eon proposalFiles
 
+    -- Extract return addresses from proposals and check that the return address in each proposal is registered
+
+    let returnAddrHashes =
+          Set.fromList
+            [ StakeCredentialByKey returnAddrHash
+            | (proposal, _) <- proposals
+            , let (_, returnAddrHash, _) = fromProposalProcedure eon proposal -- fromProposalProcedure needs to be adjusted so that it works with script hashes.
+            ]
+        treasuryWithdrawalAddresses =
+          Set.fromList
+            [ stakeCred
+            | (proposal, _) <- proposals
+            , let (_, _, govAction) = fromProposalProcedure eon proposal
+            , TreasuryWithdrawal withdrawalsList _ <- [govAction] -- Match on TreasuryWithdrawal action
+            , (_, stakeCred, _) <- withdrawalsList -- Extract fund-receiving stake credentials
+            ]
+        allAddrHashes = Set.union returnAddrHashes treasuryWithdrawalAddresses
```

Comment:
Maybe it would be nicer to pattern-match only once on each proposal. I would make a `getStakeAddressesFromProposal` function that returns a `Set` and just `mconcatMap` over it. Maybe a `getStakeAddressFromGovAction` inside. But this is fine, just an idea, up to you.